liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,420 @@
+            .expect("The file must not end before an EOI.");
+        if element.as_rule() == Rule::EOI {
+            return panic!("Errors not implemented. Unclosed block.");
+        }
+        if element.as_rule() == Rule::Tag {
+            let nested_tag_name = element
+                .clone() // Maybe there is a better way?
+                .into_inner()
+                .next()
+                .expect("Tags start by their identifier.")
+                .as_str();
+            if self.name == nested_tag_name {
+                self.nesting_depth += 1;
+            } else if self.end_name == nested_tag_name {
+                self.nesting_depth -= 1;
+            }
+            if self.nesting_depth == 0 {
+                return Ok(None);
+            }
+        }
+        Ok(Some(element))
     }
 
-    #[test]
-    fn object_dot_access() {
-        let tokens = granularize("post.number").unwrap();
-        let mut context = Context::new();
-        let mut post = Object::new();
-        post.insert("number".into(), Value::scalar(42i32));
-        context.stack_mut().set_global("post", Value::Object(post));
-
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+    fn close(mut self) -> Result<()> {
+        while let Some(_) = self.next()? {}
+        Ok(())
     }
 
-    #[test]
-    fn object_index_access() {
-        let tokens = granularize("post[\"number\"]").unwrap();
-        let mut context = Context::new();
-        let mut post = Object::new();
-        post.insert("number".into(), Value::scalar(42i32));
-        context.stack_mut().set_global("post", Value::Object(post));
-
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+    pub fn parse(&mut self, options: &LiquidOptions) -> Result<Vec<Box<Renderable>>> {
+        let mut renderables = Vec::new();
+        while let Some(element) = self.next()? {
+            renderables.push(parse_element(element, self.iter, options)?);
+        }
+        Ok(renderables)
     }
 
-    #[test]
-    fn object_variable_access() {
-        let tokens = granularize("post[foo]").unwrap();
-        let mut context = Context::new();
-        let mut post = Object::new();
-        post.insert("number".into(), Value::scalar(42i32));
-        context.stack_mut().set_global("post", Value::Object(post));
-        context
-            .stack_mut()
-            .set_global("foo", Value::scalar("number"));
-
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+    pub fn to_string(&mut self) -> Result<String> {
+        let mut s = String::new();
+        while let Some(element) = self.next()? {
+            s.push_str(element.as_str());
```

Comment:
This is used by `raw`, right?

What if we have
```liquid
{raw}
{raw}
{endraw}
```

Won't the use of `next` cause this to fail?

Reply:
Yep. On the current parser, though, the problem is even bigger.
Not only this also happens (tries to nest raw tags) but it doesn't even require to close the block.
The output is like this:
`{%raw%}{%endraw%}`
Which is worse because not only it nests them, but it also allows the blocks to be freely unclosed.

I didn't bother much with this case because `{%endraw%}` will already be unable to be used normally inside a `raw` block no matter what. I assumed writting `{%raw%}` inside `raw` blocks also wouldn't be needed in a practical scenario. If you prefer, I can try to create a way for blocks to block nesting.

Reply:
If it is an existing problem, I recommend creating an issue and moving on.

Reply:
Thanks!